### PR TITLE
Fusible Sensors

### PR DESCRIPTION
**Pull request by oldgalileo**

### CODE REVIEW

File: src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package org.usfirst.frc4904.sovereignty;
+
+
+import org.usfirst.frc4904.robot.vision.AligningCamera;
+import edu.wpi.first.wpilibj.PIDSource;
+import edu.wpi.first.wpilibj.PIDSourceType;
+
+public class AlignmentSystem implements PIDSource {
+	public enum AlignmentState {
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java]: Make static

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
+package org.usfirst.frc4904.sovereignty;
+
+
+public class FusedSensor<T> implements Fusible<T> {
```
Comment (carturn) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: While I understand the theoretical motivations for using a generic here, are there any cases we have encountered that use anything except a double?

File: src/org/usfirst/frc4904/sovereignty/FusibleNavX.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	@Override
+	public Double getValue() {
+		return getAngle() - relativeOffset;
+	}
+
+	@Override
+	public boolean trustable() {
+		return true;
```
Comment (carturn) [inline src/org/usfirst/frc4904/sovereignty/FusibleNavX.java]: The NavX has an isConnected function, which might be useful here.

**Comment by carturn** [comment]
I think it would be better not to name this system Fusible. In a lot of contexts, I think "sensor fusion" tends to imply a combination of multiple sensor values simultaneously to increase accuracy (e.g. combining encoder data with NavX data to get more accurate location data). I would suggest something related to failover or fallback.
I am definitely very much in favour of this system as a whole though; I think it might solve a lot of problems.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+		this(primarySensor, fallbackSensor, null);
+	}
+
+	protected Fusible<T> getActiveSensor() {
+		if (primarySensor.trustable()) {
+			return primarySensor;
+		} else if (fallbackSensor.trustable()) {
+			return fallbackSensor;
+		}
+		return lastResortSensor;
```
Comment (carturn) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java:70]: When lastResortSensor is used might be a decent time to throw an InvalidSensorException.

File: src/org/usfirst/frc4904/robot/vision/AligningCamera.java [hunk 5]
```diff
@@ -26,16 +27,16 @@ public AligningCamera() {
...
 
-	public float getDistance() {
-		return (float) cameraTable.getNumber(AligningCamera.FIELD_DISTANCE, Float.NaN);
+	public double getDistance() {
+		return cameraTable.getNumber(AligningCamera.FIELD_DISTANCE, Double.NaN);
 	}
 
 	public boolean isVisible() {
-		return cameraTable.getBoolean(AligningCamera.FIELD_VISIBLE, getDegrees() == Float.NaN || getDistance() == Float.NaN);
+		return cameraTable.getBoolean(AligningCamera.FIELD_VISIBLE, getDegrees() == Double.NaN || getDistance() == Double.NaN);
```
Comment (leijurv) [inline src/org/usfirst/frc4904/robot/vision/AligningCamera.java]: Note that in Java, *nothing* is == to NaN, not even Double.NaN itself. `Double.NaN == Double.NaN` results in `false`.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 2 above)
Comment (leijurv) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: MAYBE it could be an INT or a FLOAT

Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: As we talked about, for extensibility and usability it makes sense to keep this as a generic and leave the output type up to the implementing sensor. While attempting to demonstrate my solution to Yoni, which initially used doubles, I chose to implement a button as a use case. It became immediately clear that the output should be variable to prevent unnecessary manual casting or conversion.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	public T getValueDangerously() throws InvalidSensorException {
+		T value = getActiveSensor().getValue();
+		if (value != null) {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Invert this if statement?

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 4 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:70]: I implemented this in getValueDangerously instead.

**Review by ajnadel — APPROVED** [review]

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+		return value;
+	}
+
+	@Override
+	public T getValue() {
+		T value = null;
+		try {
+			value = getValueDangerously();
+		}
+		catch (InvalidSensorException ex) {}
```
Comment (leijurv) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: this is safe

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 4 above)
Comment (carturn) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:70]: I am not sure `getValueDangerously` is a good naming scheme to start using; I think that it implies that it should not be used, which is not what we want to imply.

File: src/org/usfirst/frc4904/sovereignty/FusibleNavX.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package org.usfirst.frc4904.sovereignty;
+
+
+import org.usfirst.frc4904.standard.custom.sensors.NavX;
+import edu.wpi.first.wpilibj.SerialPort.Port;
+
+public class FusibleNavX extends NavX implements Fusible<Double> {
```
Comment (carturn) [inline src/org/usfirst/frc4904/sovereignty/FusibleNavX.java:7]: I do not think it would be problematic to add the Fusible interface directly to the NavX; "NavX" is already a subclass we made mostly in order to implement various custom interfaces.

**Review by carturn — APPROVED** [review]
Other than naming issues, I approve. Please test. Make sure to look for edge cases.

File: .classpath [hunk 9]
```diff
@@ -5,6 +5,10 @@
 	<classpathentry kind="var" path="networktables" sourcepath="networktables.sources"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/navx_frc.jar" sourcepath="USERLIBS_DIR/navx_frc-sources.jar"/>
-	<classpathentry kind="var" path="USERLIBS_DIR/CTRLib.jar"/>
+        <classpathentry kind="var" path="USERLIBS_DIR/CTRLib.jar" sourcepath="USERLIBS_DIR/CTRLib-javadoc.jar"/>
```
Comment (c0d3rman) [inline .classpath]: Are these changes necessary? I think we should merge the opencv / cscore ones so they stop showing up on every PR but the other ones seem redundant.

File: src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	public void cancel() {
+		state = AlignmentState.IDLE;
+	}
+
+	public double getDegrees() {
+		try {
+			return sensor.getValueDangerously();
+		}
+		catch (InvalidSensorException ex) {
+			LogKitten.wtf("AlignmentSystem: no sensors were available.");
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java]: This should probably not be a WTF and should also log the exception.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
+package org.usfirst.frc4904.sovereignty;
+
+
+import org.usfirst.frc4904.standard.custom.sensors.InvalidSensorException;
+
+public class FusedSensor<T> implements Fusible<T> {
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: I agree with Carter - this might better be named `FailoverSensor` or something similar.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+		this.fallbackSensor = lastSensor;
+		this.lastResortSensor = new Fusible<T>() {
+			@Override
+			public T getValue() {
+				return defaultValue;
+			}
+
+			@Override
+			public boolean trustable() {
+				return false;
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java:29]: Shouldn't this always be `true`?

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+public class FusedSensor<T> implements Fusible<T> {
+	protected final Fusible<T> primarySensor;
+	protected final Fusible<T> fallbackSensor;
+	protected final Fusible<T> lastResortSensor;
+
+	public FusedSensor(T defaultValue, Fusible<T> primarySensor, Fusible<T> secondarySensor, Fusible<T>... sensors) {
+		Fusible<T> lastSensor = sensors[sensors.length - 1];
+		for (int i = sensors.length - 2; i >= 0; i--) {
+			lastSensor = new FusedSensor<>(sensors[i], lastSensor);
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Why nest these as opposed to simply storing the array? It seems like the logic would be mostly the same but it would be easier to understand/debug.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+		}
+		lastSensor = new FusedSensor<>(secondarySensor, lastSensor);
+		this.primarySensor = primarySensor;
+		this.fallbackSensor = lastSensor;
+		this.lastResortSensor = new Fusible<T>() {
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java:21]: Instead of duplicating this logic, can we call the other constructor(s)?

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+			@Override
+			public T getValue() {
+				return defaultValue;
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java:24]: So this is just a default value for the sensor? Perhaps we should make a class for it in a separate file.

Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:24]: Also reflect it in the naming - it's not really a "sensor".

File: src/org/usfirst/frc4904/sovereignty/FusibleNavX.java (hunk 8 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusibleNavX.java:7]: I agree. I added it like this for the sake of not making a branch on Standard. That's why it's in Sovereignty (all the stuff I put in here is standard quality for the most part and doesn't depend on anything in the yearly code.)

**Comment by oldgalileo** [comment]
I agree on the naming front but couldn't decide on anything better. I'm hoping Osher got around to testing but if not I'll see to making that happen on Thursday.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+	protected Fusible<T> getActiveSensor() {
+		if (primarySensor.trustable()) {
+			return primarySensor;
+		} else if (fallbackSensor.trustable()) {
+			return fallbackSensor;
+		}
+		return lastResortSensor;
+	}
+
+	public T getValueDangerously() throws InvalidSensorException {
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: As Carter said, this should probably match our `getSafely` convention

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+		T value = getActiveSensor().getValue();
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: This will throw a null pointer exception if the third constructor (which sets the `lastResortSensor` to `null`) is used and all sensors fail.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+		if (value == null) {
+			throw new InvalidSensorException();
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Perhaps include a message to indicate which sensor failed?

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 7 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Agreed - some good error logging should be done here. This is the automatic error handling for people who are too lazy to do it themselves, so it should be the best generalized handling we can muster.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+		T value = null;
+		try {
+			value = getValueDangerously();
+		}
+		catch (InvalidSensorException ex) {}
+		return value;
+	}
+
+	@Override
+	public boolean trustable() {
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java:95]: I don't think this is the best name, but I can't come up with a better one. Suggestions?

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+		if (primarySensor.trustable() || fallbackSensor.trustable()) {
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Won't this mean that the fallbackSensor will always be trustable if any of the non-primary sensors are? Since this is called recursively? This is one reason I think it would be better to use an array.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 13 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Also, here we construct lastSensors iteratively but below we use them recursively - we should do one or the other.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 20 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Also why not just `return primarySensor.trustable() || fallbackSensor.trustable()`? Why the if statement?

File: src/org/usfirst/frc4904/sovereignty/FusibleNavX.java (hunk 8 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusibleNavX.java:7]: I think this is fine for now, since the NavX is a large conglomeration of sensors and any one of them could be returned in `getValue()`. In mid-season we will generalize this.

File: src/org/usfirst/frc4904/sovereignty/FusibleNavX.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+	public boolean trustable() {
+		return isMagnetometerCalibrated();
+	}
+
+	public void zeroRelativeOffset() {
+		setRelativeOffset(getAngle());
+	}
+
+	public void setRelativeOffset(double offset) {
+		relativeOffset = offset;
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusibleNavX.java]: Could the parameter be renamed such that this is `this.relativeOffset = relativeOffset`?

File: src/org/usfirst/frc4904/sovereignty/strategies/GearAlign.java [hunk 22]
```diff
@@ -16,14 +18,22 @@
...
-	protected AligningCamera camera;
+	protected AlignmentSystem alignSystem;
 	protected TrimmablePIDController pidController;
 	protected boolean onAngle = false;
 	private ChassisMove chassisMove;
 
-	public GearAlign(AligningCamera camera) {
-		this.camera = camera;
-		pidController = new TrimmablePIDController(GearAlign.ANGLE_P, GearAlign.ANGLE_I, GearAlign.ANGLE_D, camera);
+	public GearAlign() {
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/strategies/GearAlign.java:24]: These cascading constructors are pretty pointless for yearly code - let's go with just this one and move all functionality into it.

**Review by c0d3rman — CHANGES_REQUESTED** [review]
Very good system. I've suggested some cleanup / refactoring and pointed out some bugs. We will generalize this system further and document it in the mid/offseason.

File: .classpath (hunk 9 above)
Comment (oldgalileo) [reply .classpath]: Uh, this change adds the CTRlib-javadoc which we don't have by default. I can open a seperate PR for that but it should be done at some point. It's annoying to not be able to fetch docs on stuff.

File: src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java (hunk 10 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java]: I forgot that non-wtf errors are a thing. That being said, in this case it is accurate since both of these not working IS a terrible failure.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 11 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Failover's not a bad suggestion. If you want me to go with that I can.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 15 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:24]: This is a backup for if all the provided sensors are down and you want to provide a default value instead of null and an exception. It's treated in code as a default sensor just for simplicity, but having it as it's own file is kinda unnecessary.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 12 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:29]: No, because this value is your last resort and trustable should reflect confidence in the accuracy of the result of which this is literally the most opposite thing from that in this scenario.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 4 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:70]: Alternative? I was using it for its property of being the antonym of safely. In this case, since we are overriding the getValue method it can't throw an error, but an error should undoubtedly be thrown here as you commented so to resolve that I went with the dangerous route.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 16 above)
Comment (HowardStark) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: See comment above for reason why that isn't possible in this scenario.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 18 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: They both failed here?

File: .classpath (hunk 9 above)
Comment (carturn) [reply .classpath]: I do not seem to have a CTRLib-javadoc.jar, so this is probably not required and will cause errors on some setups.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 19 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:95]: I think it reflects the functionality here quite accurately. We are asking if the data the sensor is providing is trustable. This could be through checking if the sensor is accurate/online, or perhaps if the data has normalized. As such, it's not a simple state of "off/on" but whether the data is acceptable for use.

File: src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java (hunk 10 above)
Comment (carturn) [reply src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java]: Good catch. Exception log is used elsewhere for this.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 12 above)
Comment (carturn) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:29]: Should not it be an or of the sensors?

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 20 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: That's exactly what you would want, though. In a system with, say, 6 sensors, the system should be trustable if you can get real data from any of the sensors. So, if the 6th or lowest priority sensor was trustable, in the root sensor the fallback sensor would return true. And within that sensor, the fallback sensor would return true, and so on and so forth, thus allowing for trustability to propagate through the entire chain.

File: src/org/usfirst/frc4904/sovereignty/FusibleNavX.java (hunk 8 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusibleNavX.java:7]: Sure.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 4 above)
Comment (carturn) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:70]: Is there a reason that Fusible does not simply have getValue and getValueSafely?
This has been setup for other standard interfaces as well.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 13 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: I'm not sure I understand your comment on doing this recursively in other places, but doing this recursively would be pretty inefficient. You would have to call System.arrayCopy or Array.copyOf on each call of the constructor to shorten the sensors array. With this system, it's pretty efficient, and it saves us from having to throw errors in the constructor or anything. Maybe I'm missing something, idk.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 18 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Well this would happen when whichever sensor was active returned a null value. This wouldn't happen with the last resort sensor. But it could happen if a sensor reported it was trustable but returned null. (Do we want to allow that?)

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 12 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:29]: I thought trustable wasn't reflecting confidence in the sensor, but whether the sensor was communicating + good enough to be used. The last resort sensor never goes offline. (And the paradigm for all other sensors is "never return a value from a non-`trustable()` sensor)

File: src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java (hunk 10 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/AlignmentSystem.java]: WTF is only really used for debugging. No sensors being available sucks, but it is not even `FATAL` and may not even be an `ERROR` (we might want it as a `WARN` although I think error is probably better)

File: .classpath (hunk 9 above)
Comment (c0d3rman) [reply .classpath]: I was annoyed by the lack of CTRlib javadoc too and started adding it, but right now it's not in 4904-USERLIBS so we can't really include it here. Also a couple lines down the NavX javadoc was readded in spite of already being on line 7.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 11 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Go for it!

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 4 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:70]: I think that for ease of integration we should definitely keep only `getValue` in `Fusible`, and if we want `getValueSafely` we should have it in a child interface.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 20 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Oh, I see. I wasn't thinking about it correctly and thought it would simply always default to the second sensor in the chain.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 14 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:21]: No. You have to build the recursive sensors before setting that stuff, and you have to make the call to another constructor as the first thing you do. (Unless of course you want me to jank it up)

File: src/org/usfirst/frc4904/sovereignty/FusibleNavX.java (hunk 21 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusibleNavX.java]: ooh nice.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 18 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Oooh. That's crazy man. That's craaaaazy. Okay. Will fix. (No, but we should handle that. Can't make people do everything perfectly (unfortunately))

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 13 above)
Comment (c0d3rman) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: OK, that makes sense. But my original question still stands - why do this recursively at all? It seems just iterating over an array and calling `trustable` would be simpler and easier.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 12 above)
Comment (carturn) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:29]: Oops, did not look at enough context. I am inclined to agree that trustable here should return true, as this sensor is trustable for its intended purpose. However, any sensors which "fuse" with a lastResortSensor should not pass on this trust to higher levels.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+		this(null, primarySensor, fallbackSensor);
+	}
+
+	protected Fusible<T> getActiveSensor() {
+		if (primarySensor.trustable()) {
+			return primarySensor;
+		} else if (fallbackSensor.trustable()) {
+			return fallbackSensor;
+		}
+		return lastResortSensor;
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/sovereignty/FusedSensor.java:70]: Does every nested FusibleSensor maintain a fallbackSensor? Because it seems like only the outer one would ever be used.

Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:70]: Not really. If the default value isn't passed in, it's just null.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 24]
```diff
@@ -89,18 +87,20 @@
...
 
 	public RobotMap() {
 		Component.pdp = new PDP();
 		Component.shifter = new SolenoidShifters(Port.Pneumatics.solenoidUp, Port.Pneumatics.solenoidDown);
-		Component.navx = new NavX(SerialPort.Port.kMXP);
-		Component.chassisTurnMC = new CustomPIDController(0.01, 0.0, -0.02, RobotMap.Component.navx);
-		Component.chassisTurnMC.setInputRange(-180, 180);
-		Component.chassisTurnMC.setContinuous(true);
+		Component.navx = new FusibleNavX(SerialPort.Port.kMXP);
+		Component.chassisDriveMC = new CustomPIDController(0.01, 0.0, -0.02, RobotMap.Component.navx);
```
Comment (c0d3rman) [inline src/org/usfirst/frc4904/robot/RobotMap.java:101]: Why the name change? This is still a Turn MC.

File: .classpath (hunk 9 above)
Comment (ajnadel) [reply .classpath]: The second NavX entry is some kind of equilibrium that is reached in Eclipse. It will happen after a certain amount of time.

File: src/org/usfirst/frc4904/robot/RobotMap.java (hunk 24 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/robot/RobotMap.java:101]: I feel like I didn't do that.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 12 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java:29]: @c0d3rman well, you kinda hit on either side of the target. It's really a mixture of both of those. The status of the sensor, i.e. communicating, isn't enough. You also need to know if the data is okay for use, which might not be the case despite having comms for a plethora of reasons. The sensor can most likely determine for itself how accurate the data is. If it is, say, roughly normal it would probably return `true`. So `trustable` is really the confidence the sensor has in itself. 

In this particular case with the lastResortSensor, I would argue that the application of that confidence or perhaps the scope of that confidence is nuanced. In the FusedSensor, lastResortSensor is only used when both the primary and fallback sensor are un-trustable. At this point, it makes sense to deem any further value passed from the FusedSensor un-trustable, as they are not from any of the sensors provided. The last resort sensor is simply a courtesy to the user intended to allow them to handle the failing of both sensors in a more elegant way than catching a null pointer. It is not, however, meant to actually be acted upon ever, and I would thus deem that value un-trustable.

File: src/org/usfirst/frc4904/sovereignty/FusedSensor.java (hunk 13 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: No particular reason, really. I just liked this implementation.

Comment (oldgalileo) [reply src/org/usfirst/frc4904/sovereignty/FusedSensor.java]: Actually, I guess partly I didn't like the idea of littering the file unnecessarily with for loops. One on construction vs. one each time a value is fetched. Idk.

**Comment by leijurv** [comment]
screen shot 2017-02-11 at 1 24 33 pm
This is the 69th comment on this pull request

File: .classpath [hunk 25]
```diff
@@ -5,7 +5,11 @@
 	<classpathentry kind="var" path="networktables" sourcepath="networktables.sources"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/navx_frc.jar" sourcepath="USERLIBS_DIR/navx_frc-sources.jar"/>
-        <classpathentry kind="var" path="USERLIBS_DIR/CTRLib.jar" sourcepath="USERLIBS_DIR/CTRLib-javadoc.jar"/>
+	<classpathentry kind="var" path="USERLIBS_DIR/CTRLib.jar">
```
Comment (c0d3rman) [inline .classpath]: These are good changes but should definitely happen in another branch
